### ISSUE TITLE
ThreadSanitizer issues

### ISSUE THREAD

**Issue by sadiqj** [ocaml/ocaml#11040]
There's been discussion around ThreadSanitizer runs of trunk in multiple venues but it's probably best to concentrate it in one place so things don't get lost.

To begin with, I did a ThreadSanitizer build of [a88c1dbe0a17164e0afc3eb8fbc75c9d867c9992](https://github.com/ocaml/ocaml/tree/a88c1dbe0a17164e0afc3eb8fbc75c9d867c9992) ran tests/parallel from the testsuite. This gives the following issues:

- [x] atomics.ml - race in `caml_darken` (https://gist.github.com/sadiqj/b4da4d6663e66dcef4fe008c3f8b9e09)
- [x] atomics.ml - race in `mark_slice_darken` (https://gist.github.com/sadiqj/2050aaa11ef9df79e0558264dba7cb6f)
- [x] atomics.ml - race in `oldify_one` (https://gist.github.com/sadiqj/efd3ba388de78ac9aeac79654312ac44)
- [x] domain_parallel_spawn_burn.ml - race in `mark_slice_darken` (with `mark_slice_darken` this time) (https://gist.github.com/sadiqj/64edd37cd9c3b0beeaa6c7a7c3a8f81e)
- [x] domain_parallel_spawn_burn.ml - race in `camel_darken` (https://gist.github.com/sadiqj/5b9aafc280d800a76de377f0e8ab8c5b)
- [x] finaliser_handover.ml - race in `pool_global_adopt` (https://gist.github.com/sadiqj/19e8ab1dcc825bd66d7510f5c65f6d17)
- [x] finaliser_handover.ml - race in `caml_add_orphaned_finalisers` and `no_orphaned_work` (https://gist.github.com/sadiqj/87cba99f11e9c805461a2bcc4bf95a98)

I'm deduplicating by the pair of racing callsites. This isn't exact though (the second callsite can be "some previous access") so there may be duplicates.

Unfortunately there's no threading support in issues so things will start to get messy pretty quick. If it looks like there are bigger chunks of discussion on issues it probably makes sense to create a new issue for them (and add them to the tasks above they concern).

---

*Edit from @gasche*: another list from @OlivierNicole [below](https://github.com/ocaml/ocaml/issues/11040#issuecomment-1438855993) is as follows:

- [x] `oldify_one` / `caml_empty_minor_heap_promote` in `lib-format/mc_pr586_par2.ml` https://gist.github.com/OlivierNicole/b287312779cbe46565258a89fb002d56
- [x] `caml_natdynlink_open` / `oldify_one` in `lib-dynlink-domains/main.ml` https://gist.github.com/OlivierNicole/c0d29f4dd50936943c965f90eb8dadb3
- [x] `caml_empty_minor_heap_promote` / `try_update_object_header` in `lib-format/mc_pr586_par.ml` https://gist.github.com/OlivierNicole/37d9bb3efbb71e255a392f6d2ad5f0c3
- [x] `caml_empty_minor_heap_promote` / `oldify_one` in the same
- [x] `try_update_object_header` / `do_some_marking` in `lib-runtime-events/test_caml_slot_reuse.ml` https://gist.github.com/OlivierNicole/673c606f87456450820d3a7bdbfdd17d
- [x] `caml_accum_alloc_stats` / `caml_collect_alloc_stats_sample` in `parallel/major_gc_wait_backup.ml` https://gist.github.com/OlivierNicole/ba7822bce91f56b0c7b404f742c91282
- [x] `caml_accum_heap_stats` / `caml_collect_heap_stats_sample` in the same
- [x] `mark_stack_push_block` / `caml_darken` in `parallel/mctest.ml` https://gist.github.com/OlivierNicole/3a79cbc988983ba294ea44f8e10d3477
- [x] `do_some_marking` / `caml_modify` in `parallel/multicore_systhreads.ml` https://gist.github.com/OlivierNicole/232e634bb29ceac7f0734e6fb793fd0b
- [x] `no_orphaned_work` / `caml_adopt_orphaned_work` in `weak-ephe-final/finaliser_handover` https://gist.github.com/OlivierNicole/e88a42c02dcbc4a7e5f064bbe6d7cb26
- [x] `pool_initialize` / `Hd_val` in `lib-runtime-events/test_caml_parallel` https://gist.github.com/OlivierNicole/b39776b97bdb115397119f36ee92d3ec
- [x] `compute_gc_stats` / `caml_collect_heap_stats_sample` in `parallel/major_gc_wait_backup` https://gist.github.com/OlivierNicole/9ba2a2dd4252586505f493677850721b
- [x] `Hd_val` / `caml_shared_try_alloc` in `parallel/mctest` https://gist.github.com/OlivierNicole/d3444ce27fe270b0eb43b8166637281e

**Comment by sadiqj** [comment]
The PR https://github.com/ocaml/ocaml/pull/11110 adds a selection of `CAMLno_tsan` attributes and relaxed atomics. It also adds an ignorelist for functions where the races reported are from `Field` and `Tag_val` macros.

There's a broader question here. I'm not sure we're realistically going to be able to run ThreadSanitizer clean without investing a fair amount of work in changing the runtime and I'm wondering if the payoff is worth it. `Field` and `Tag_val` are l-values and so we can't do the same thing we did with `Hd_val`. To get around this we'd have to introduce a parallel set of macros to use for loading/storing field and tag vaues (which could maybe only be exposed internally to begin with?). We'd then have to move the runtime's use of `Field` and `Tag_val` over.

There's a discussion going in https://github.com/ocaml/ocaml/issues/10992 for `Field` that could end up going down this path.

I've a few more failures to whittle away at but they're mostly adding annotations. A common remaining issue is that we have some fast paths in our atomics when there's only a single domain[^fast_load]. So ThreadSanitizer can see the non-atomic fast paths as races with the atomics when there's eventually more than one domain.

[^fast_load]: I have to question how valuable the one in https://github.com/ocaml/ocaml/blob/trunk/runtime/memory.c#L249 is.

**Comment by kayceesrk** [comment]
> A common remaining issue is that we have some fast paths in our atomics when there's only a single one

Would making them relaxed atomics fix the issue?

**Comment by sadiqj** [comment]
> Would making them relaxed atomics fix the issue?

Yes, I think that would work.

**Comment by gasche** [comment]
> we have some fast paths in our atomics when there's only a single one. So ThreadSanitizer can see the non-atomic fast paths as races with the atomics when there's eventually more than one domain.

At first I thought that there cannot be a race here: if the domain is alone, then only that domain can spawn new domains, and so non-atomic `if (caml_domain_alone())` sections that do not spawn new threads cannot be racy. But in fact this is not strictly guaranteed, as in theory those functions could be called from a multi-threaded C program, with another thread creating a new domain (on the same OCaml runtime) in parallel, or possibly from a signal that would preempt the thread in the middle of the `if (caml_domain_alone())` section. In these implausible situations the implementation is buggy, in the sense that it does not actually guarantee atomicity.

Using relaxed atomcis would silence the warning but would not prevent the loss of atomicity. To guarantee atomicity, a critical-section mechanism would need to be devised, for example replacing the `load` on `caml_num_domains_running` in the test by a CAS from `1` to `0`, and having additional-domain-creating functions wait in this situation. This sounds expensive and complex, so probably not a good idea.

**Comment by xavierleroy** [comment]
@maranget and I wondered about this single-domain special case in the implementation of the Atomic.t operations: how much of a speedup does it give w.r.t. always using the C11 atomic operations?  Just curious.

**Comment by xavierleroy** [comment]
> There's a broader question here. I'm not sure we're realistically going to be able to run ThreadSanitizer clean without investing a fair amount of work in changing the runtime and I'm wondering if the payoff is worth it.

Good question!  In my opinion, the usefulness of running ThreadSanitizer is to find race conditions that we did not expect, e.g. a missing lock somewhere.  But to see these (hypothetical) race conditions we need first to silence all the alarms corresponding to expected race conditions.  And this may be too much work (and too much noise added to the runtime system, and too many risks of introducing new bugs) for the expected benefit.

**Comment by sadiqj** [comment]
> At first I thought that there cannot be a race here: if the domain is alone, then only that domain can spawn new domains, and so non-atomic `if (caml_domain_alone())` sections that do not spawn new threads cannot be racy. But in fact this is not strictly guaranteed, as in theory those functions could be called from a multi-threaded C program, with another thread creating a new domain (on the same OCaml runtime) in parallel, or possibly from a signal that would preempt the thread in the middle of the `if (caml_domain_alone())` section. In these implausible situations the implementation is buggy, in the sense that it does not actually guarantee atomicity.

I think `caml_domain_spawn` can only be called while holding the runtime lock (this applies to `caml_atomic_*` too) so I'm not sure it's possible to create a new domain in parallel.

We may call `caml_record_signal` in the middle of a fast path but this doesn't run any OCaml code. That only happens when we hit the next polling/allocation point - outside the fast path.

**Comment by gasche** [comment]
That's an excellent, point, but do `caml_atomic_*` operations require the runtime lock? This is natural for atomics that need the write barrier, but I wouldn't have necessarily expected it from `caml_atomic_fetch_add`, and it is not documented. If this really is the argument, we should clarify this.

**Comment by sadiqj** [comment]
For the fast paths to be safe there's an assumption that the runtime lock is held. e.g https://github.com/ocaml/ocaml/blob/trunk/runtime/memory.c#L226 breaks horribly in a single-runtime multiple-thread situation.

It's a good point though, if we keep the fast paths we should document this.

**Comment by gadmm** [comment]
If you do not hold the runtime lock, could there not be a race with the GC moving the block from minor to major, making the atomic operate on a stale value? The functions operate on `value`s. If you were to root the values, you would need the runtime lock; if you avoid rooting by reasoning on allocation points, then you are also reasoning that the runtime lock is held.

I agree that the fast path is safe, and the signature of the function is enough to say that they assume the runtime lock is held.

**Comment by gasche** [comment]
That makes a lot of sense indeed, sorry for the noise.

**Comment by gasche** [comment]
> `Field` and `Tag_val` are l-values and so we can't do the same thing we did with `Hd_val`. To get around this we'd have to introduce a parallel set of macros to use for loading/storing field and tag vaues (which could maybe only be exposed internally to begin with?). We'd then have to move the runtime's use of `Field` and `Tag_val` over.

One proposal: we could have a header file that *overrides* `Field` and `Tag_val` to be rvalue-only, and also defines `Store_field` and `Store_tag_val` for storage. This would limit the required changes to lvalues-only usage, which are much less common than rvalue-only usage. (For `Tag_val`, I find only two uses for storing in `runtime/*.c`; `Field` has quite a bit more.)

**Comment by xavierleroy** [comment]
FYI: I removed Thread Sanitizer from the Jenkins CI, as this discussion shows that OCaml 5 will not pass it any time soon.  "How I learned to stop worrying and love the data races"...

**Comment by OlivierNicole** [comment]
While working on [TSan for OCaml programs](https://github.com/ocaml-multicore/ocaml-tsan) with @fabbing we found a number of additional races, which we have deduplicated by pairs of callsites like @sadiqj and list below.

These were found by running the test suite on git revision https://github.com/ocaml-multicore/ocaml-tsan/tree/backup_tsan_patch.

TSan shouldn’t produce false positives, so in principle all of these should be addressed for correctness. Some of the fixes may be as simple as using a relaxed atomic access instead of a plain access.

- [x] `oldify_one` / `caml_empty_minor_heap_promote` in `lib-format/mc_pr586_par2.ml` https://gist.github.com/OlivierNicole/b287312779cbe46565258a89fb002d56
- [x] `caml_natdynlink_open` / `oldify_one` in `lib-dynlink-domains/main.ml` https://gist.github.com/OlivierNicole/c0d29f4dd50936943c965f90eb8dadb3
- [x] `caml_empty_minor_heap_promote` / `try_update_object_header` in `lib-format/mc_pr586_par.ml` https://gist.github.com/OlivierNicole/37d9bb3efbb71e255a392f6d2ad5f0c3
- [x] `caml_empty_minor_heap_promote` / `oldify_one` in the same
- [x] `try_update_object_header` / `do_some_marking` in `lib-runtime-events/test_caml_slot_reuse.ml` https://gist.github.com/OlivierNicole/673c606f87456450820d3a7bdbfdd17d
- [x] `caml_accum_alloc_stats` / `caml_collect_alloc_stats_sample` in `parallel/major_gc_wait_backup.ml` https://gist.github.com/OlivierNicole/ba7822bce91f56b0c7b404f742c91282
- [x] `caml_accum_heap_stats` / `caml_collect_heap_stats_sample` in the same
- [x] `mark_stack_push_block` / `caml_darken` in `parallel/mctest.ml` https://gist.github.com/OlivierNicole/3a79cbc988983ba294ea44f8e10d3477
- [x] `do_some_marking` / `caml_modify` in `parallel/multicore_systhreads.ml` https://gist.github.com/OlivierNicole/232e634bb29ceac7f0734e6fb793fd0b
- [x] `no_orphaned_work` / `caml_adopt_orphaned_work` in `weak-ephe-final/finaliser_handover` https://gist.github.com/OlivierNicole/e88a42c02dcbc4a7e5f064bbe6d7cb26
- [x] `pool_initialize` / `Hd_val` in `lib-runtime-events/test_caml_parallel` https://gist.github.com/OlivierNicole/b39776b97bdb115397119f36ee92d3ec
- [x] `compute_gc_stats` / `caml_collect_heap_stats_sample` in `parallel/major_gc_wait_backup` https://gist.github.com/OlivierNicole/9ba2a2dd4252586505f493677850721b
- [x] `Hd_val` / `caml_shared_try_alloc` in `parallel/mctest` https://gist.github.com/OlivierNicole/d3444ce27fe270b0eb43b8166637281e

**Comment by gasche** [comment]
I looked at one of those races, and it is really a concurrency bug in the runtime (I propose a fix in #12591), which in fact covers two different items listed here (`no_orphaned_work` appears in one bullet point by Sadiq and one by Olivier, it is the same bug). Another group of reports here (`collect_*_stats_sample`) correspond to another bug which is now discussed in #12590 and will be fixed soon.

I think we should spend more work looking at these issues, there are probably other small concurrency bugs lurking. (But then, most of the other remaining issues look a bit scary: if they involve `Hd_val` or `oldify_one` they are in performance-critical areas and deciding what to do is less easy.)

**Comment by eutro** [comment]
Here's another one distinct from, but related to, the `collect_*_stats_sample` races, and without a TSan report yet.

- [x] `caml_clear_gc_stats_sample` from `domain_terminate` races with `caml_accum_alloc_stats` in a test I wrote for it https://gist.github.com/eutro/0942ec358aaf14ff879f1ad1952daa16

**Comment by gasche** [comment]
I filled more boxes in this issue after merging #12591 (fixes races with `no_orphaned_work`) and #12595 (fixes races involving `caml_collect_gc_stats_sample`). There is another PR in discussion, #12597, which would fix the race involving `caml_clear_gc_stats_sample`.

**Comment by OlivierNicole** [comment]
I have ticked two more boxes after the merge of https://github.com/ocaml/ocaml/pull/12030 (a while ago) and the removal of the corresponding silencing annotations by https://github.com/ocaml/ocaml/pull/12693.

**Comment by OlivierNicole** [comment]
For better readability of the events timeline in this issue, let me signal that ThreadSanitizer runs have been re-enabled in Jenkins by https://github.com/ocaml/ocaml/pull/12644.

**Comment by OlivierNicole** [comment]
For the record (to avoid working concurrently on the same races), I believe I have a diagnostic and a fix for races nos. 1, 2, 3, 4 and 5 of [my post above](https://github.com/ocaml/ocaml/issues/11040#issuecomment-1438855993), and I am investigating the race reported by Sadiq in `pool_global_adopt`.

Edit.: the numbers were wrong.

**Comment by OlivierNicole** [comment]
@sadiqj I was looking at data race report no. 3 in your initial message. I think it is fixed now that `Field` performs a volatile read: we don’t consider volatile accesses to race with volatile atomics. So I think this one can be ticked.

**Comment by OlivierNicole** [comment]
I ticked box no. 5 in my own message. The race has been fixed by #12030.

**Comment by fabbing** [comment]
We can also check the box for the race on `pool_global_adopt` from your message @sadiqj since #12755!

**Comment by sadiqj** [comment]
Updated @fabbing !

**Comment by OlivierNicole** [comment]
Marking races 1 to 4 in [my post](https://github.com/ocaml/ocaml/issues/11040#issuecomment-1438855993) as fixed (see #12737).

**Comment by OlivierNicole** [comment]
The last two reports listed in this issue have been fixed by #12894. Could I suggest therefore to close this issue?

Sometimes an unknown or new data race reports comes up (e.g. #12905), but we no longer have a quantity that requires a dedicated summary issue. And hopefully this will happen less and less often!

**Comment by OlivierNicole** [comment]
I took the liberty of checking the last two boxes of this issue in accordance with my previous message. Now closing as all the reports have been addressed. (Hurray!)
